File: src/algorithms/arrays/kadane-subarray/max-product-subarray/sources/max-product-subarray_test.rs
include!("max-product-subarray.rs");

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_default_input() {
        let (max_product, _, _) = max_product_subarray(&[2, 3, -2, 4, -1, 2]);
        assert_eq!(max_product, 96);
    }

    #[test]
    fn test_all_positive() {
        let (max_product, _, _) = max_product_subarray(&[1, 2, 3, 4]);
        assert_eq!(max_product, 24);
    }

    #[test]
    fn test_with_zero() {
        let (max_product, _, _) = max_product_subarray(&[2, 3, 0, 4, 5]);
        assert_eq!(max_product, 20);
    }

    #[test]
    fn test_single_element() {
        let (max_product, start, end) = max_product_subarray(&[7]);
        assert_eq!(max_product, 7);
        assert_eq!(start, 0);
        assert_eq!(end, 0);
    }

    #[test]
    fn test_two_negatives() {
        let (max_product, _, _) = max_product_subarray(&[-2, -3]);
        assert_eq!(max_product, 6);
    }

    #[test]
    fn test_negative_flip() {
        let (max_product, _, _) = max_product_subarray(&[-2, 3, -4]);
        assert_eq!(max_product, 24);
    }

    #[test]
    fn test_empty_array() {
        let (max_product, _, _) = max_product_subarray(&[]);
        assert_eq!(max_product, 0);
    }

    #[test]
    fn test_valid_indices() {
        let input_array = [2, 3, -2, 4, -1, 2];
        let (_, start, end) = max_product_subarray(&input_array);
        assert!(start <= end);
        assert!(end < input_array.len());
    }
}

File: src/algorithms/arrays/kadane-subarray/max-product-subarray/__tests__/max-product-subarray_test.py
import importlib
import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "sources"))

module = importlib.import_module("max-product-subarray")
max_product_subarray = module.max_product_subarray


def test_default_input():
    result = max_product_subarray([2, 3, -2, 4, -1, 2])
    assert result["max_product"] == 96, f"Expected 96, got {result['max_product']}"


def test_all_positive():
    result = max_product_subarray([1, 2, 3, 4])
    assert result["max_product"] == 24


def test_with_zero():
    result = max_product_subarray([2, 3, 0, 4, 5])
    assert result["max_product"] == 20


def test_single_element():
    result = max_product_subarray([7])
    assert result["max_product"] == 7
    assert result["start_index"] == 0
    assert result["end_index"] == 0


def test_two_negatives():
    result = max_product_subarray([-2, -3])
    assert result["max_product"] == 6


def test_negative_flip():
    result = max_product_subarray([-2, 3, -4])
    assert result["max_product"] == 24


def test_empty_array():
    result = max_product_subarray([])
    assert result["max_product"] == 0


def test_valid_indices():
    input_array = [2, 3, -2, 4, -1, 2]
    result = max_product_subarray(input_array)
    assert result["start_index"] >= 0
    assert result["end_index"] < len(input_array)
    assert result["start_index"] <= result["end_index"]


if __name__ == "__main__":
    test_default_input()
    test_all_positive()
    test_with_zero()
    test_single_element()
    test_two_negatives()
    test_negative_flip()
    test_empty_array()
    test_valid_indices()
    print("All tests passed!")
